Initial checkin of source files to sb-signing-utils project

subras-ibm opened the pull request.

subras-ibm committed:
Initial checkin of source files to sb-signing-utils project

fspbld commented:
Just a general comment - can we get the signtool binary name changed so it isn't confused with the Linux-provided tool (/usr/bin/signtool)? 

SignImage example script is in general out of date with the latest signtool parms  (ie, -fld-name is actually --fldname and -fld-value is --fldvalue).
I am wondering if this script is even needed? If we are shipping my crtSignedContainer.pl tool, couldn't users just look at the calls in it (and thus save us from having to maintain 2 different scripts in this repo?)

---

Review status: 0 of 28 files reviewed at latest revision, all discussions resolved.

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/open-power/sb-signing-utils/1#-:-KQ6xEQsgby5-yNvF3-B:b-8bfe7h)_

<!-- Sent from Reviewable.io -->

subras-ibm commented:
Pam,

  Yes the SignImage under scrpts directory can be removed. 

   Nick and I talked about changing the name signtool to something else, 
but agreed that we can do this after our 9/1 delivery. BTW changing the 
name should be very trivial.

Regards,

Subramanian Swaminathan
Software Developer, LTC Security
subras@us.ibm.com
(919)254-8249   T/L 444-8249

From:   Pam Eggler notifications@github.com
To:     open-power/sb-signing-utils sb-signing-utils@noreply.github.com
Cc:     Subramanian Swaminathan1/Raleigh/IBM@IBMUS, Author 
author@noreply.github.com
Date:   08/26/2016 01:55 PM
Subject:        Re: [open-power/sb-signing-utils] Initial checkin of 
source files to sb-signing-utils project (#1)

Just a general comment - can we get the signtool binary name changed so it 
isn't confused with the Linux-provided tool (/usr/bin/signtool)? 
SignImage example script is in general out of date with the latest 
signtool parms (ie, -fld-name is actually --fldname and -fld-value is 
--fldvalue).
I am wondering if this script is even needed? If we are shipping my 
crtSignedContainer.pl tool, couldn't users just look at the calls in it 
(and thus save us from having to maintain 2 different scripts in this 
repo?)

Review status: 0 of 28 files reviewed at latest revision, all discussions 
resolved.

Comments from Reviewable
—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub, or mute the thread.

williamspatrick commented:
@subras-ibm - There is a lot of content in this repository that was generated by 'autoreconf -i'.  You should not check that content into the raw github repository.

Please clean this up to just be 'Makefile.am' and 'configure.ac'.  It is fairly common practice to also include a 'bootstrap.sh' script that has a call to 'autoreconf -i'.

williamspatrick commented:
I updated previous comment to say 'Makefile.am' and not 'Makefile.in'.  am = automake, which is the original source.

williamspatrick commented on src/signtool/Makefile.am:
```diff
@@ -0,0 +1,43 @@
+        IBM_Crypto.cpp  \
+        signtool.cpp
+
+signtool_CPPFLAGS = -I. -I -g3
```
I would strongly suggest we use C++11 or C++14 for all new C++ code.  Leaving off -std=c++11 means it compiles with c++98 (or maybe c++03).

williamspatrick commented:
I have not reviewed all of this yet because it is very difficult to navigate in github with all of the autotools generated stuff in place.  I'll review this again when it is condensed down to just your contributions.

subras-ibm committed:
Removed all autogenerated files, added bootstrap.sh script to regenerate all needed config files

subras-ibm commented:
Hi Patrick,
I have removed all autogenerated files, added bootstrap.sh to call autoreconf -i, to regenerate all needed files and added c++11 flags to Makefile.am

bofferdn commented:
Review status: 0 of 18 files reviewed at latest revision, 46 unresolved discussions.

---

_[bootstrap.sh, line 1 [r2]](https://reviewable.io:443/reviews/open-power/sb-signing-utils/1#-KRFruNypnYyzGX5N5mw:-KRFruNypnYyzGX5N5mx:b9w6w3j) ([raw file](https://github.com/open-power/sb-signing-utils/blob/3ee191b598680985fce744380aba0f9282d67faf/bootstrap.sh#L1)):_

> ``` Shell
> #!/bin/sh
> ```

Does this need our standard prolog?  Or is this copied and not our original work?

---

_[configure.ac, line 1 [r2]](https://reviewable.io:443/reviews/open-power/sb-signing-utils/1#-KRFrhPhvob9U1oGu38x:-KRFrhPioeoI2dx1Vfr7:b-sonylo) ([raw file](https://github.com/open-power/sb-signing-utils/blob/3ee191b598680985fce744380aba0f9282d67faf/configure.ac#L1)):_

> ``` M4Sugar
> #-------------------------------------------------------------------------
> ```

We should have the standard header I pointed to, with the right dates, project names, license, etc.

---

_[configure.ac, line 13 [r2]](https://reviewable.io:443/reviews/open-power/sb-signing-utils/1#-KRFs89NZSvGspr27kmM:-KRFs89NZSvGspr27kmN:b-khfl4e) ([raw file](https://github.com/open-power/sb-signing-utils/blob/3ee191b598680985fce744380aba0f9282d67faf/configure.ac#L13)):_

> ``` M4Sugar
> # GSA ADP Schedule Contract with IBM Corporation.
> #---------------------------------------------------------------------------
> AC_INIT(SecureAndTrustedBoot, 1.0.0, subras@us.ibm.com)
> ```

is this suppose to be the package name?  Should it say SecurebootSigningUtils ?

---

_[configure.ac, line 15 [r2]](https://reviewable.io:443/reviews/open-power/sb-signing-utils/1#-KRFsg6sLSiqqbEzwdd2:-KRFsg6sLSiqqbEzwdd3:b2luyt1) ([raw file](https://github.com/open-power/sb-signing-utils/blob/3ee191b598680985fce744380aba0f9282d67faf/configure.ac#L15)):_

> ``` M4Sugar
> AC_INIT(SecureAndTrustedBoot, 1.0.0, subras@us.ibm.com)
> AC_CONFIG_AUX_DIR([config])
> AM_INIT_AUTOMAKE([-Wall -Werror foreign -Wno-portability])
> ```

have we tried compiling this on POWER architecture at all?  so we can drink our own koolaid?

---

_[configure.ac, line 33 [r2]](https://reviewable.io:443/reviews/open-power/sb-signing-utils/1#-KRFsQ-Uixy1J7f8DRRm:-KRFsQ-Uixy1J7f8DRRn:b-vg7bps) ([raw file](https://github.com/open-power/sb-signing-utils/blob/3ee191b598680985fce744380aba0f9282d67faf/configure.ac#L33)):_

> ``` M4Sugar
> 
> # Checks for typedefs, structures, and compiler characteristics.
>  #AC_C_CONST
> ```

is it standard to have all these options commented out (throughout config)?  Any reason to keep them?

---

_[Makefile.am, line 1 [r2]](https://reviewable.io:443/reviews/open-power/sb-signing-utils/1#-KRFt-6pR_DrkK8tfi65:-KRFt-6pR_DrkK8tfi66:bg8t1ed) ([raw file](https://github.com/open-power/sb-signing-utils/blob/3ee191b598680985fce744380aba0f9282d67faf/Makefile.am#L1)):_

> ``` Makefile
> #/* IBM_PROLOG_BEGIN_TAG                                                   */
> ```

for all prologs, we don't need the /\* ... */ on files that use # as comment indicator

---

_[README.md, line 1 [r2]](https://reviewable.io:443/reviews/open-power/sb-signing-utils/1#-KRFt9nAY06zxEVfDPIq:-KRFt9nAY06zxEVfDPIr:bon0ycl) ([raw file](https://github.com/open-power/sb-signing-utils/blob/3ee191b598680985fce744380aba0f9282d67faf/README.md#L1)):_

> ``` Markdown
> To sign an official build, we need sign_agent to talk to the signing server via Notes. 
> ```

We should strip all references to lotus Notes here and below?  I thought we were going with a more open solution.

---

_[scripts/SignImage, line 1 [r2]](https://reviewable.io:443/reviews/open-power/sb-signing-utils/1#-KRFtTaW4MrLcJjvMmRi:-KRFtTaW4MrLcJjvMmRj:buf88oz) ([raw file](https://github.com/open-power/sb-signing-utils/blob/3ee191b598680985fce744380aba0f9282d67faf/scripts/SignImage#L1)):_

> ```
> #!/bin/bash -x
> ```

prolog?

---

_[src/Makefile.am, line 1 [r2]](https://reviewable.io:443/reviews/open-power/sb-signing-utils/1#-KRFtgOV3yLI4-mIfOVi:-KRFtgOV3yLI4-mIfOVj:b-7b0jap) ([raw file](https://github.com/open-power/sb-signing-utils/blob/3ee191b598680985fce744380aba0f9282d67faf/src/Makefile.am#L1)):_

> ``` Makefile
> #/* IBM_PROLOG_BEGIN_TAG                                                   */
> ```

Just # , not /\* ... */

---

_[src/signtool/IBM_Container.cpp, line 75 [r2]](https://reviewable.io:443/reviews/open-power/sb-signing-utils/1#-KRFw-F5Q8Jv0WKjYRwW:-KRFw-F6gMCQk5Cc6ziT:b-grwbg4) ([raw file](https://github.com/open-power/sb-signing-utils/blob/3ee191b598680985fce744380aba0f9282d67faf/src/signtool/IBM_Container.cpp#L75)):_

> ``` cpp
>         uint64_t value = 0;
> 
>         value = (            data[0]        | ((uint16_t)data[1] << 8)  | 
> ```

seems like an algorithm would be less cumbersome here

---

_[src/signtool/IBM_Container.cpp, line 83 [r2]](https://reviewable.io:443/reviews/open-power/sb-signing-utils/1#-KRFw62s2hnq3w94cazp:-KRFw62s2hnq3w94cazq:bgld5hp) ([raw file](https://github.com/open-power/sb-signing-utils/blob/3ee191b598680985fce744380aba0f9282d67faf/src/signtool/IBM_Container.cpp#L83)):_

> ``` cpp
>     }
> 
>     void printHexBytes( const uint8_t* bytes, uint32_t size, uint32_t leadSpace )
> ```

unsigned sized (like size + leadSpace) are often better represented by size_t

---

_[src/signtool/IBM_Container.cpp, line 87 [r2]](https://reviewable.io:443/reviews/open-power/sb-signing-utils/1#-KRFuqnmhDxOB3DCWSgE:-KRFuqnmhDxOB3DCWSgF:b-k8kokh) ([raw file](https://github.com/open-power/sb-signing-utils/blob/3ee191b598680985fce744380aba0f9282d67faf/src/signtool/IBM_Container.cpp#L87)):_

> ``` cpp
>         std::vector<uint8_t> byteArray;
> 
>         for (uint32_t i = 0; i < size; i++)
> ```

possible to use the st::vector.assign() ?
like   byteArray.assign(bytes,bytes+size) ?

---

_[src/signtool/IBM_Container.cpp, line 105 [r2]](https://reviewable.io:443/reviews/open-power/sb-signing-utils/1#-KRFvBO2hGNyOuCm2MCq:-KRFvBO2hGNyOuCm2MCr:beqqdn1) ([raw file](https://github.com/open-power/sb-signing-utils/blob/3ee191b598680985fce744380aba0f9282d67faf/src/signtool/IBM_Container.cpp#L105)):_

> ``` cpp
>         static std::map<int, std::string> s_hashAlgoNames;
> 
>         if (s_hashAlgoNames.size() == 0)
> ```

s_hashAlgoNames.empty() is usually a less costly check

---

_[src/signtool/IBM_Container.cpp, line 107 [r2]](https://reviewable.io:443/reviews/open-power/sb-signing-utils/1#-KRFwRVEvzonAX7qIX29:-KRFwRVF8KWRJ046Wk2n:b-tevt51) ([raw file](https://github.com/open-power/sb-signing-utils/blob/3ee191b598680985fce744380aba0f9282d67faf/src/signtool/IBM_Container.cpp#L107)):_

> ``` cpp
>         if (s_hashAlgoNames.size() == 0)
>         {
>             s_hashAlgoNames[HASH_ALG_SHA512] = "SHA-512";
> ```

symbolic constants for SHA-512?

---

_[src/signtool/IBM_Container.cpp, line 110 [r2]](https://reviewable.io:443/reviews/open-power/sb-signing-utils/1#-KRFvLqS5BOo9hBH28ie:-KRFvLqS5BOo9hBH28if:bf4nxk) ([raw file](https://github.com/open-power/sb-signing-utils/blob/3ee191b598680985fce744380aba0f9282d67faf/src/signtool/IBM_Container.cpp#L110)):_

> ``` cpp
>         }
> 
>         std::map<int, std::string>::iterator itr = s_hashAlgoNames.find(p_hashAlgoNum);
> ```

const_iterator instead of iterator

---

_[src/signtool/IBM_Container.cpp, line 113 [r2]](https://reviewable.io:443/reviews/open-power/sb-signing-utils/1#-KRFvT2Hbvoh4ScH1g1w:-KRFvT2Hbvoh4ScH1g1x:b-cc9tbx) ([raw file](https://github.com/open-power/sb-signing-utils/blob/3ee191b598680985fce744380aba0f9282d67faf/src/signtool/IBM_Container.cpp#L113)):_

> ``` cpp
>         if (itr != s_hashAlgoNames.end())
>         {
>             return itr->second.c_str();
> ```

We;ve often had a guideline of no return from middle of function, even though it can make things simpler.  This is a general comment for all code; we could discuss if needed.

---

_[src/signtool/IBM_Container.cpp, line 124 [r2]](https://reviewable.io:443/reviews/open-power/sb-signing-utils/1#-KRFvdBOZoKD7RNtqV50:-KRFvdBOZoKD7RNtqV51:b75txuo) ([raw file](https://github.com/open-power/sb-signing-utils/blob/3ee191b598680985fce744380aba0f9282d67faf/src/signtool/IBM_Container.cpp#L124)):_

> ``` cpp
>         static std::map<int, std::string> s_signAlgoNames;
> 
>         if (s_signAlgoNames.size() == 0)
> ```

.empty() instead of .size()

---

_[src/signtool/IBM_Container.cpp, line 126 [r2]](https://reviewable.io:443/reviews/open-power/sb-signing-utils/1#-KRFwYxpvbryDIkf5UFX:-KRFwYxpvbryDIkf5UFY:b-2mc134) ([raw file](https://github.com/open-power/sb-signing-utils/blob/3ee191b598680985fce744380aba0f9282d67faf/src/signtool/IBM_Container.cpp#L126)):_

> ``` cpp
>         if (s_signAlgoNames.size() == 0)
>         {
>             s_signAlgoNames[SIG_ALG_ECDSA521] = "ECDSA521";
> ```

symbolic names?

---

_[src/signtool/IBM_Container.cpp, line 129 [r2]](https://reviewable.io:443/reviews/open-power/sb-signing-utils/1#-KRFvhlMSS0qyUG4kk_p:-KRFvhlMSS0qyUG4kk_q:b-lf1ib5) ([raw file](https://github.com/open-power/sb-signing-utils/blob/3ee191b598680985fce744380aba0f9282d67faf/src/signtool/IBM_Container.cpp#L129)):_

> ``` cpp
>         }
> 
>         std::map<int, std::string>::iterator itr = s_signAlgoNames.find((int) p_signAlgoNum);
> ```

const_iterator

---

_[src/signtool/IBM_Container.cpp, line 139 [r2]](https://reviewable.io:443/reviews/open-power/sb-signing-utils/1#-KRFweL2xO8S34FqOPSu:-KRFweL2xO8S34FqOPSv:b-4kk5fd) ([raw file](https://github.com/open-power/sb-signing-utils/blob/3ee191b598680985fce744380aba0f9282d67faf/src/signtool/IBM_Container.cpp#L139)):_

> ``` cpp
> 
> 
>     void ReadPublicKeyFromFile( std::string           p_mode,
> ```

pass strings as const std:string& ?

const method?

---

_[src/signtool/IBM_Container.cpp, line 142 [r2]](https://reviewable.io:443/reviews/open-power/sb-signing-utils/1#-KRFxgXrkKitA7j_ZhdW:-KRFxgXrkKitA7j_ZhdX:b-8vdyyx) ([raw file](https://github.com/open-power/sb-signing-utils/blob/3ee191b598680985fce744380aba0f9282d67faf/src/signtool/IBM_Container.cpp#L142)):_

> ``` cpp
>                                 std::string           p_keyFileName,
>                                 std::vector<uint8_t>& p_buffer,
>                                 int                   p_size )
> ```

any pointer to specifying a size?  It seems like this API already has to be intimately aware that it is 132 bytes + 1 prefix byte.  Why can't the caller just say get the public key from file X into this buffer.  Then this routine asks for 133 bytes initially, remove 1 byte, and return a 132 byte buffer to caller?

---

_[src/signtool/IBM_Container.cpp, line 146 [r2]](https://reviewable.io:443/reviews/open-power/sb-signing-utils/1#-KRFwwceLSa6Ft8edJco:-KRFwwceLSa6Ft8edJcp:bpj85kk) ([raw file](https://github.com/open-power/sb-signing-utils/blob/3ee191b598680985fce744380aba0f9282d67faf/src/signtool/IBM_Container.cpp#L146)):_

> ``` cpp
>         IBM_Utils* pUtils = IBM_Utils::get();
>         THROW_EXCEPTION(pUtils == NULL);
> ```

should p_buffer be cleared here?  What if there is content in it already?

---

_[src/signtool/IBM_Container.cpp, line 154 [r2]](https://reviewable.io:443/reviews/open-power/sb-signing-utils/1#-KRFxHaWc4kfmF9uFKNT:-KRFxHaWc4kfmF9uFKNU:b-sujmku) ([raw file](https://github.com/open-power/sb-signing-utils/blob/3ee191b598680985fce744380aba0f9282d67faf/src/signtool/IBM_Container.cpp#L154)):_

> ``` cpp
>                                                 // at the begining
> 
>             if (p_buffer[0] != 0x04)
> ```

symbolic constant?

---

_[src/signtool/IBM_Container.cpp, line 175 [r2]](https://reviewable.io:443/reviews/open-power/sb-signing-utils/1#-KRFyTbKUs_DfoOPCuYl:-KRFyTbKUs_DfoOPCuYm:be8gpy6) ([raw file](https://github.com/open-power/sb-signing-utils/blob/3ee191b598680985fce744380aba0f9282d67faf/src/signtool/IBM_Container.cpp#L175)):_

> ``` cpp
>         {
>             // should never get here
>             THROW_EXCEPTION_STR("Invalid mode.");
> ```

better error msg?  Like  print out the p_mode and indicate must be the dev or prod keyword?

---

_[src/signtool/IBM_Container.cpp, line 182 [r2]](https://reviewable.io:443/reviews/open-power/sb-signing-utils/1#-KRFy33S1cmBZ7nJAtI0:-KRFy33TwBjpLil3a93o:b18rg1y) ([raw file](https://github.com/open-power/sb-signing-utils/blob/3ee191b598680985fce744380aba0f9282d67faf/src/signtool/IBM_Container.cpp#L182)):_

> ``` cpp
> 
> 
>     void ReadSignatureFromFile( std::string           p_mode,
> ```

const std::string& passing 

const method?

---

_[src/signtool/IBM_Container.cpp, line 186 [r2]](https://reviewable.io:443/reviews/open-power/sb-signing-utils/1#-KRFy5nfyz2Hebqu4j4y:-KRFy5nfyz2Hebqu4j4z:b-9gwocn) ([raw file](https://github.com/open-power/sb-signing-utils/blob/3ee191b598680985fce744380aba0f9282d67faf/src/signtool/IBM_Container.cpp#L186)):_

> ``` cpp
>                                 std::vector<uint8_t>& p_buffer,
>                                 int                   p_size )
>     {
> ```

same question on size; does the caller really have to compute the size ahead of time, or can this function just know it ?

---

_[src/signtool/IBM_Container.cpp, line 188 [r2]](https://reviewable.io:443/reviews/open-power/sb-signing-utils/1#-KRFyv47Yr9-bsSo7Jif:-KRFyv47Yr9-bsSo7Jig:b-gam4c8) ([raw file](https://github.com/open-power/sb-signing-utils/blob/3ee191b598680985fce744380aba0f9282d67faf/src/signtool/IBM_Container.cpp#L188)):_

> ``` cpp
>     {
>         IBM_Utils* pUtils = IBM_Utils::get();
>         THROW_EXCEPTION(pUtils == NULL);
> ```

should p_buffer be clear here?

---

_[src/signtool/IBM_Container.cpp, line 201 [r2]](https://reviewable.io:443/reviews/open-power/sb-signing-utils/1#-KRFye4-ZgoQko-ck0ui:-KRFye4-ZgoQko-ck0uj:b-pypnu4) ([raw file](https://github.com/open-power/sb-signing-utils/blob/3ee191b598680985fce744380aba0f9282d67faf/src/signtool/IBM_Container.cpp#L201)):_

> ``` cpp
>         {
>             // should never get here
>             THROW_EXCEPTION_STR("Invalid mode.");
> ```

better error message, like comment above?

---

_[src/signtool/IBM_Container.cpp, line 229 [r2]](https://reviewable.io:443/reviews/open-power/sb-signing-utils/1#-KRH6lvicFd2-IUybwhL:-KRH6lvicFd2-IUybwhM:bk8jx7g) ([raw file](https://github.com/open-power/sb-signing-utils/blob/3ee191b598680985fce744380aba0f9282d67faf/src/signtool/IBM_Container.cpp#L229)):_

> ``` cpp
>  * @param[out]  packet  -  A reference to a vector of type uint8_t
>  */
> void ContainerHdr::GetHeaderBytes( std::vector<uint8_t>& packet )
> ```

const method?

---

_[src/signtool/IBM_Container.cpp, line 239 [r2]](https://reviewable.io:443/reviews/open-power/sb-signing-utils/1#-KRH4AFnSvv2KiM88eu4:-KRH4AFnSvv2KiM88eu5:b-82piif) ([raw file](https://github.com/open-power/sb-signing-utils/blob/3ee191b598680985fce744380aba0f9282d67faf/src/signtool/IBM_Container.cpp#L239)):_

> ``` cpp
>     packet.push_back((magicNumber >> 8) & 0xFF);
>     packet.push_back((magicNumber >> 16) & 0xFF);
>     packet.push_back((magicNumber >> 24) & 0xFF);
> ```

not quite getting this .. magic number starts as little endian, then you convert to big endian, then back to little endian for the insertions?  why not just skip the double conversion?  
Could you also use  c++11 insert like   packet.insert (packet.end(), reinterpret_cast<uint8_t*>(magicNumber),reinterpret_cast<uint8_t*>(magicNumber),+sizeof(magicNumber) );

Same general comment for all the handling below as well.

Worst case, some function that reverses its inputs in place, so that a standard insert will work here?

---

_[src/signtool/IBM_Container.cpp, line 302 [r2]](https://reviewable.io:443/reviews/open-power/sb-signing-utils/1#-KRH7IUEIqCAqnlLOy1e:-KRH7IUEIqCAqnlLOy1f:bk8jx7g) ([raw file](https://github.com/open-power/sb-signing-utils/blob/3ee191b598680985fce744380aba0f9282d67faf/src/signtool/IBM_Container.cpp#L302)):_

> ``` cpp
> /**
>  *  @brief   print the Container Header
>  */
> ```

const method?

---

_[src/signtool/IBM_Container.cpp, line 326 [r2]](https://reviewable.io:443/reviews/open-power/sb-signing-utils/1#-KRH7ZAff-kVyRcD9wqi:-KRH7ZAff-kVyRcD9wqj:bq8p5ml) ([raw file](https://github.com/open-power/sb-signing-utils/blob/3ee191b598680985fce744380aba0f9282d67faf/src/signtool/IBM_Container.cpp#L326)):_

> ``` cpp
> 
>     std::cout << "   m_hwPkeyA          = ";
>     printHexBytes( m_hwPkeyA, ECDSA521_KEY_SIZE, 24 );
> ```

aren't these three 132 bytes?  is 24 just an arbitrary subset to get a good feel as to what the value is?  Shouldn't this value (24) be symbolic constant?

---

_[src/signtool/IBM_Container.cpp, line 359 [r2]](https://reviewable.io:443/reviews/open-power/sb-signing-utils/1#-KRH7rSz5TjaSvZ0G1Fl:-KRH7rSz5TjaSvZ0G1Fm:bk8jx7g) ([raw file](https://github.com/open-power/sb-signing-utils/blob/3ee191b598680985fce744380aba0f9282d67faf/src/signtool/IBM_Container.cpp#L359)):_

> ``` cpp
>  * @param[out]  packet  -  A reference to a vector of type uint8_t
>  */
> void PrefixHdr::GetHeaderBytes( std::vector<uint8_t>& packet )
> ```

const method?

---

_[src/signtool/IBM_Container.cpp, line 361 [r2]](https://reviewable.io:443/reviews/open-power/sb-signing-utils/1#-KRH81K15lcEmcP5U9nh:-KRH81K15lcEmcP5U9ni:bk02jp8) ([raw file](https://github.com/open-power/sb-signing-utils/blob/3ee191b598680985fce744380aba0f9282d67faf/src/signtool/IBM_Container.cpp#L361)):_

> ``` cpp
> void PrefixHdr::GetHeaderBytes( std::vector<uint8_t>& packet )
> {
>     packet.clear();
> ```

same general comments as above w/ regard to big/little endian  

---

_[src/signtool/IBM_Container.cpp, line 437 [r2]](https://reviewable.io:443/reviews/open-power/sb-signing-utils/1#-KRH8F9evaqX3-0gShEt:-KRH8F9evaqX3-0gShEu:b-3xh1dp) ([raw file](https://github.com/open-power/sb-signing-utils/blob/3ee191b598680985fce744380aba0f9282d67faf/src/signtool/IBM_Container.cpp#L437)):_

> ``` cpp
>  *  @brief   print the Prefix Header
>  */
> void PrefixHdr::PrintHeader()
> ```

const method

---

_[src/signtool/IBM_Container.cpp, line 498 [r2]](https://reviewable.io:443/reviews/open-power/sb-signing-utils/1#-KRH8bm3d3BwYx9et6wv:-KRH8bm4GLVWNHZCgFMx:b-3xh1dp) ([raw file](https://github.com/open-power/sb-signing-utils/blob/3ee191b598680985fce744380aba0f9282d67faf/src/signtool/IBM_Container.cpp#L498)):_

> ``` cpp
>  * @param[out]  packet  -  A reference to a vector of type uint8_t
>  */
> void PrefixData::GetHeaderBytes( std::vector<uint8_t>& packet )
> ```

const method

---

_[src/signtool/IBM_Container.cpp, line 550 [r2]](https://reviewable.io:443/reviews/open-power/sb-signing-utils/1#-KRH8mG4VWwSNeSIKgYT:-KRH8mG4VWwSNeSIKgYU:b-3xh1dp) ([raw file](https://github.com/open-power/sb-signing-utils/blob/3ee191b598680985fce744380aba0f9282d67faf/src/signtool/IBM_Container.cpp#L550)):_

> ``` cpp
>  *  @brief   print the Prefix Data
>  */
> void PrefixData::PrintHeader()
> ```

const method

---

_[src/signtool/IBM_Container.cpp, line 583 [r2]](https://reviewable.io:443/reviews/open-power/sb-signing-utils/1#-KRH8uICNH2BjI-07lZu:-KRH8uICNH2BjI-07lZv:b-3xh1dp) ([raw file](https://github.com/open-power/sb-signing-utils/blob/3ee191b598680985fce744380aba0f9282d67faf/src/signtool/IBM_Container.cpp#L583)):_

> ``` cpp
>  *  @brief   Get Number of valid Software keys
>  */
> int PrefixData::GetSwKeyCount()
> ```

const method

---

_[src/signtool/IBM_Utils.cpp, line 66 [r1]](https://reviewable.io:443/reviews/open-power/sb-signing-utils/1#-KQEBKyUT-APHVZkuhv8:-KQEBKyUT-APHVZkuhv9:b-3v4l35) ([raw file](https://github.com/open-power/sb-signing-utils/blob/7b3b74844fadcb38a24a63f33eb7160f0f574179/src/signtool/IBM_Utils.cpp#L66)):_

> ``` cpp
> 
> 
> void IBM_Utils::ReadFromFile( const char*        p_fileName,
> ```

why not pass in as const string& ?

---

_[src/signtool/IBM_Utils.cpp, line 67 [r1]](https://reviewable.io:443/reviews/open-power/sb-signing-utils/1#-KQEBH72rlUD_FMvawlq:-KQEBH72rlUD_FMvawlr:b-3xh1dp) ([raw file](https://github.com/open-power/sb-signing-utils/blob/7b3b74844fadcb38a24a63f33eb7160f0f574179/src/signtool/IBM_Utils.cpp#L67)):_

> ``` cpp
> 
> void IBM_Utils::ReadFromFile( const char*        p_fileName,
>                               std::vector<byte>& p_buffer )
> ```

const method

---

_[src/signtool/IBM_Utils.cpp, line 79 [r1]](https://reviewable.io:443/reviews/open-power/sb-signing-utils/1#-KQEMOnZbmpMMv1GWHM1:-KQEMOnZbmpMMv1GWHM2:bpui4jt) ([raw file](https://github.com/open-power/sb-signing-utils/blob/7b3b74844fadcb38a24a63f33eb7160f0f574179/src/signtool/IBM_Utils.cpp#L79)):_

> ``` cpp
> 
>     // get length of file
>     ifs.seekg( 0, std::ios_base::end );
> ```

there are several ways the stream can become bad during these operations (seekg, tellg, read, close,etc.); do we need to check the stream state.  Do we need to call the good() function to ensure stream is still in a good state at each step?  That's a general comment across all file accesses.

ifs.tellg() can return -1  on error 

---

_[src/signtool/Makefile.am, line 1 [r2]](https://reviewable.io:443/reviews/open-power/sb-signing-utils/1#-KRTrA_KJ53rmMSSWNx2:-KRTrA_LDVsEOofO3306:bua21m1) ([raw file](https://github.com/open-power/sb-signing-utils/blob/3ee191b598680985fce744380aba0f9282d67faf/src/signtool/Makefile.am#L1)):_

> ``` Makefile
> #/* IBM_PROLOG_BEGIN_TAG                                                   */
> ```

only # .. /\* ... */ not also needed

---

_[src/signtool/signtool.cpp, line 47 [r1]](https://reviewable.io:443/reviews/open-power/sb-signing-utils/1#-KQE0TIbin0amFzbmlaU:-KQE0TIbin0amFzbmlaV:bv1antc) ([raw file](https://github.com/open-power/sb-signing-utils/blob/7b3b74844fadcb38a24a63f33eb7160f0f574179/src/signtool/signtool.cpp#L47)):_

> endiannes"
> endiannes => endianness

---

_[src/signtool/signtool.cpp, line 82 [r1]](https://reviewable.io:443/reviews/open-power/sb-signing-utils/1#-KQE10HlH_vT9Jk7SXCn:-KQE10HlH_vT9Jk7SXCo:b-gmf1xj) ([raw file](https://github.com/open-power/sb-signing-utils/blob/7b3b74844fadcb38a24a63f33eb7160f0f574179/src/signtool/signtool.cpp#L82)):_

> ``` cpp
>     static std::string  s_VERSION = "1.0.0";
> 
>     std::string getAppVersion(bool version)
> ```

I don't quite get the "version" parameter, because we're already getting a version; is a better name "verbose" ?

---

_[src/signtool/signtool.cpp, line 523 [r1]](https://reviewable.io:443/reviews/open-power/sb-signing-utils/1#-KQE3_E9ONMbzWvQ2Xul:-KQE3_E9ONMbzWvQ2Xum:bl2mp1p) ([raw file](https://github.com/open-power/sb-signing-utils/blob/7b3b74844fadcb38a24a63f33eb7160f0f574179/src/signtool/signtool.cpp#L523)):_

> ``` cpp
>             switch (status)
>             {
>                 case 1:
> ```

symbolic constants for these  status cases? (and the setting of rc below?)

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/open-power/sb-signing-utils/1)_

<!-- Sent from Reviewable.io -->

subras-ibm committed:
Fixed all review comments from Nick Bofferding

subras-ibm commented:
Fixed code with al review comments, give it atry again

subras-ibm committed:
Fixed wrongly spelt constant

bofferdn commented:
Review status: 0 of 18 files reviewed at latest revision, 50 unresolved discussions.

---

_[bootstrap.sh, line 4 at r4](https://reviewable.io:443/reviews/open-power/sb-signing-utils/1#-KRWlFF_2NG2luJw-ywf:-KRWlFF_2NG2luJw-ywg:b7gysv2) ([raw file](https://github.com/open-power/sb-signing-utils/blob/348ff131f6bd4e63ed893b34ab4b74dfbba5fc57/bootstrap.sh#L4)):_

> ``` Shell
> # This is an automatically generated prolog.
> #
> # $Source: src/Makefile.am $
> ```

wrong file path or name .. should be bootstrap.sh

---

_[bootstrap.sh, line 26 at r4](https://reviewable.io:443/reviews/open-power/sb-signing-utils/1#-KRWlLnyHzbgbLNLHH8Y:-KRWlLnyHzbgbLNLHH8Z:bo4wqpn) ([raw file](https://github.com/open-power/sb-signing-utils/blob/348ff131f6bd4e63ed893b34ab4b74dfbba5fc57/bootstrap.sh#L26)):_

> ``` Shell
> # IBM_PROLOG_END_TAG
> 
> #!/bin/sh
> ```

#!/bin/sh has to go to very top of file to execute as a shell script.

---

_[configure.ac, line 4 at r4](https://reviewable.io:443/reviews/open-power/sb-signing-utils/1#-KRWlX546tATypolEdLW:-KRWlX55lB-I0np9fxBW:b-olefr3) ([raw file](https://github.com/open-power/sb-signing-utils/blob/348ff131f6bd4e63ed893b34ab4b74dfbba5fc57/configure.ac#L4)):_

> ``` M4Sugar
> # This is an automatically generated prolog.
> #
> # $Source: src/Makefile.am $
> ```

wrong file path or name  .. should be configure.ac

---

_[Makefile.am, line 1 at r2](https://reviewable.io:443/reviews/open-power/sb-signing-utils/1#-KRFt-6pR_DrkK8tfi65:-KRWm7vcgls29YUnAcj5:b32w0x) ([raw file](https://github.com/open-power/sb-signing-utils/blob/3ee191b598680985fce744380aba0f9282d67faf/Makefile.am#L1)):_
<details><summary>

<i>Previously, bofferdn (Nick Bofferding) wrote…</i></summary>



> for all prologs, we don't need the /\* ... */ on files that use # as comment indicator
> </details>
> OK

---

_[scripts/SignImage, line 6 at r4](https://reviewable.io:443/reviews/open-power/sb-signing-utils/1#-KRWmapoJ8jBb2G3AHZE:-KRWmapoJ8jBb2G3AHZF:b5v8s9j) ([raw file](https://github.com/open-power/sb-signing-utils/blob/348ff131f6bd4e63ed893b34ab4b74dfbba5fc57/scripts/SignImage#L6)):_

> ```
> # This is an automatically generated prolog.
> #
> # $Source: src/Makefile.am $
> ```

File or path wrong .. should be scripts/SignImage

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/open-power/sb-signing-utils/1)_

<!-- Sent from Reviewable.io -->

subras-ibm committed:
Fixed file name in scripts/SignImage, configure.ac and bootstrap.sh
Moved #!/bin/sh to beginning of the file bootstrap.sh

subras-ibm committed:
Changed c++11 flag to c++0x until we have a newer toolchain

williamspatrick commented on configure.ac:
```diff
@@ -0,0 +1,119 @@
+AC_PROG_CXX
+
+# Checks for libraries.
+ # FIXME: Replace `main' with a function in `-lrt':
```
FIXME needs to be resolved.

williamspatrick commented on configure.ac:
```diff
@@ -0,0 +1,119 @@
+
+
+# Check host for build type
+AM_CONDITIONAL(BLD_X86,  [test x${host}  = x] )
```
Why is this referencing x86?  We do builds on Power as well.

williamspatrick commented on configure.ac:
```diff
@@ -0,0 +1,119 @@
+   src/signtool/Makefile
+])
+
+AC_OUTPUT
```
The next files is "docs/AutomatedSigining.odt".  We strongly encourage documentation to not be in a binary format, like ODT, but instead in Markdown or RST format.  ODT files cannot be version controlled.

williamspatrick commented on scripts/SignImage:
```diff
@@ -0,0 +1,114 @@
+ 
+SIGNTOOL=./signtool
+
+IMG_FILE=./images/skiboot.bin
```
I'm unsure what the purpose of this tool is, but should all of these be ?= set?  We also support XZ compressed skiboot images which are named skiboot.bin.xz.

stewart-ibm replied:
well, usually skiboot.lid and skiboot.lid.xz

and there's no reason why it *has* to be skiboot either.

williamspatrick commented on src/signtool/IBM_Container.cpp:
```diff
@@ -0,0 +1,1767 @@
+
+namespace
+{
+    const static uint32_t g_COL_WIDTH = 24;
```
`constexpr uint32_t g_COL_WIDTH` is preferred C++11 syntax.  `const static` can cause linker problems with duplicate symbol names because a const static is a run-time constant and not a compile-time constant.

williamspatrick commented on src/signtool/IBM_Container.cpp:
```diff
@@ -0,0 +1,1767 @@
+
+    const static uint32_t g_UNCOMPRESSED_PUBKEY_FORMAT = 0x04;
+
+    const static std::string g_HASH_ALGO_SHA512_NAME   = "SHA-512";
```
C++11 would tend to prefer string literal syntax.

constexpr auto g_HASH_ALGO_SHA512_NAME = "SHA-512"s;

williamspatrick commented on src/signtool/IBM_Container.cpp:
```diff
@@ -0,0 +1,1767 @@
+  
+    const char* GetHashAlgoName( int p_hashAlgoNum )
+    {
+        static std::map<int, std::string> s_hashAlgoNames;
```
This can be compile-time initialized with C++11.

static const std::map<int, std::string> s_hashAlgoNames = { { HASH_ALG_SHA512 , g_HASH_ALGO_SHA512_NAME },  { HASH_ALG_UNKNOWN, "Unknown"s } };

williamspatrick commented on src/signtool/IBM_Container.cpp:
```diff
@@ -0,0 +1,1767 @@
+            hashAlgoName = itr->second.c_str();
+        }
+
+        return hashAlgoName.c_str();
```
What is the reason for this to return a const char\* instead of a const string&?

williamspatrick commented on src/signtool/IBM_Container.cpp:
```diff
@@ -0,0 +1,1767 @@
+    };
+
+  
+    const char* GetSignAlgoName( int p_signAlgoNum )
```
Same comments on this function as previous.

williamspatrick commented on src/signtool/IBM_Container.cpp:
```diff
@@ -0,0 +1,1767 @@
+    packet.clear();
+
+    // insert the magic nummber field
+    packet.push_back((m_magicNumber >> 24) & 0xFF);
```
I suspect these could all be significantly cleaner with an insert call from a std::array or initializer_list.

```
auto getByte = [](auto value, auto byte) { return (value >> (byte*8)) & 0xff; }

packet.insert(packet.end(), { getByte(m_magicNumber, 3), getByte(m_magicNumber, 2) ... });
```

williamspatrick commented on src/signtool/IBM_Container.cpp:
```diff
@@ -0,0 +1,1767 @@
+        THROW_EXCEPTION_STR(ss.str().c_str());
+    }
+
+    int padLen = 4096;
```
size_t padLen = PAGE_SIZE;

williamspatrick commented on src/signtool/IBM_Container.cpp:
```diff
@@ -0,0 +1,1767 @@
+
+        case e_CONTAINER_SIZE:
+        {
+            uint64_t containerSize;
```
Why aren't you using size_t for sizes and counts?

williamspatrick commented on src/signtool/IBM_Container.cpp:
```diff
@@ -0,0 +1,1767 @@
+}
+
+
+void IBM_Container::initializeMap()
```
Use compile-time initializer_list initialization.

williamspatrick commented on src/signtool/IBM_Container.cpp:
```diff
@@ -0,0 +1,1767 @@
+    int keyCount = 0;
+
+    // Parse the ContainerHdr
+    m_containerHdr.m_magicNumber = getUint32( pRawData );
```
I would have done 'getUint32', very different.

```
namespace details
{
    template <typename T> struct getField
    {
        void op(T&, uint8_t**) = delete;
    };

    template <> struct getField<uint8_t>
    {
        void op(uint8_t& value, uint8_t** buffer)
        {
             value = **buffer;
             *buffer += sizeof(value);
        }
    };
    /// ... duplicate for other types, doing endianness fixups as appropriate.
}

template < typename T > void getField(T& value, uint8_t** buffer) { return details::getField<T>::op(value, buffer); }
```

Now all of these lines are just:

```
getField(m_containerHdr.m_magicNumber, pRawData);
```

or if you want to be explicit:

```
getField<uint32_t>(m_containerHdr.m_magicNumber, pRawData);
```

williamspatrick commented on src/signtool/IBM_Container.h:
```diff
@@ -0,0 +1,253 @@
+
+struct ContainerHdr
+{
+    uint32_t  m_magicNumber;                    // (17082011)
```
C++11 allows class default initializers.

uint32_t m_magicNumber = MAGIC;

This allows you to eliminate the constructor.

williamspatrick commented on src/signtool/IBM_Container.h:
```diff
@@ -0,0 +1,253 @@
+    void PrintHeader() const;
+    void GetHeaderBytes( std::vector<uint8_t>& packet ) const;
+
+}  __attribute__ ((packed));
```
What is the reason for these to be packed?  Are you doing a raw memcpy of the full object?

williamspatrick commented on src/signtool/IBM_Container.h:
```diff
@@ -0,0 +1,253 @@
+                      std::string& p_digestStr );
+
+private:
+    // Disallow Copy Constructor and Assignment Operator
```
You also need to deal with the move functions.  And the preferred way in C++11 is `=delete`, not making them private.  
Private doesn't entirely solve the problem because friends are still impacted and it causes linker-time errors instead of compile-time errors.

williamspatrick commented on src/signtool/IBM_Crypto.cpp:
```diff
@@ -0,0 +1,702 @@
+IBM_Crypto::IBM_Crypto( const std::string& p_mode )
+   : m_mode(p_mode)
+{
+    if ( !((m_mode == IBM_Utils::g_MODE_PRODUCTION) || 
```
Why isn't m_mode a specific enum that can only be in these two states?  Take care of it at compile-time instead of run-time.

williamspatrick commented on src/signtool/IBM_Utils.h:
```diff
@@ -0,0 +1,92 @@
+                            std::vector<byte>& p_buffer );
+
+private:
+    IBM_Utils();                                //!< disallow 
```
Also not the correct way.

williamspatrick commented on src/signtool/Makefile.am:
```diff
@@ -0,0 +1,43 @@
+        IBM_Crypto.cpp  \
+        signtool.cpp
+
+signtool_CPPFLAGS = -I. -I -g3 -std=c++0x
```
The -std=c++0x should be expressed in the configure.ac instead via the AX_CXX_COMPILE_STDCXX_0X macro.  This way you can support compilers other than gcc and you will get ./configure-time checking for support.

subras-ibm committed:
Updated configure.ac to check support for c++11/c++0x during configure stage
and update compiler flags appropriately

stewart-ibm commented on README.md:
```diff
@@ -0,0 +1,8 @@
+production mode.
+
+The development mode uses keys generated from an OpenSSL environment, while a production mode
+uses the IBM Crypto Card to generate and sign builds.
```
Is a *requirement* to use an IBM crypto card? Or is that just an implementation detail of how IBM will do things?

stewart-ibm commented on README.md:
```diff
@@ -0,0 +1,8 @@
+uses the IBM Crypto Card to generate and sign builds.
+
+A more detailed info about the signtool is in AutomatedSigning.odt file under the docs directory, a
+more complete example of signtool usage is in SignImage under the scripts directory.
```
It would be better to have all the documentation viewable as text.

stewart-ibm commented on README.md:
```diff
@@ -0,0 +1,8 @@
+The signtool is an utility that can be used to issue sign/verify requests for a development or
+production mode.
```
What is development mode?

What is production mode?

WHy would you do one over the other?

stewart-ibm commented on README.md:
```diff
@@ -0,0 +1,8 @@
+The signtool is an utility that can be used to issue sign/verify requests for a development or
```
```
$ which signtool
/usr/bin/signtool
```

pick a different executable name. Maybe op-signtool ?

cclaudio replied:
The op-signtool name seems to be something specific for openpower. IIRC, signtool tool will be used for openpower and powervm.

stewart-ibm commented on Makefile.am:
```diff
@@ -0,0 +1,29 @@
+# Import the c++11 detection macros in the m4 dir
+ACLOCAL_AMFLAGS = -I m4
+
+SUBDIRS = src
```
```
$ make -j
make  all-recursive
make[1]: Entering directory '/home/stewart/sb-signing-utils'
Making all in src
make[2]: Entering directory '/home/stewart/sb-signing-utils/src'
Making all in signtool
make[3]: Entering directory '/home/stewart/sb-signing-utils/src/signtool'
```

See http://aegis.sourceforge.net/auug97.pdf

stewart-ibm commented on configure.ac:
```diff
@@ -0,0 +1,115 @@
+
+# Checks for libraries.
+ # FIXME: Replace `main' with a function in `-lrt':
+ AC_CHECK_LIB([rt], [main])
```
Fix this, surely this is not the current autoconf best practice

stewart-ibm commented on configure.ac:
```diff
@@ -0,0 +1,115 @@
+AC_PROG_CXX
+
+# Checks for c++11 compatibility
+AX_CXX_COMPILE_STDCXX_11([noext],[optional])
```
I know that IBM has some truly ancient build machines (RHEL6). Does this work on RHEL6?

stewart-ibm commented on configure.ac:
```diff
@@ -0,0 +1,115 @@
+AC_CONFIG_FILES([
+   Makefile 
+   src/Makefile
+   src/signtool/Makefile
```
Again, see Recursive Make Considered Harmful.

stewart-ibm commented on configure.ac:
```diff
@@ -0,0 +1,115 @@
+#--------------------------------------------------------------------
+# compiler stuff
+#--------------------------------------------------------------------
+MY_CFLAGS="-Wall -Wno-sign-compare  -include config.h  -I \$(EXPORT_INCDIR)"
```
For something security sensitive, I'd expect to see a lot more hardening CFLAGS and CXXFLAGS by default.

stewart-ibm commented on scripts/SignImage:
```diff
@@ -0,0 +1,114 @@
+# IBM_PROLOG_BEGIN_TAG
+# This is an automatically generated prolog.
+#
+# $Source: scripts/SignImage $
```
CamelCase binary names are generaly frowned upon.

stewart-ibm commented on scripts/SignImage:
```diff
@@ -0,0 +1,114 @@
+
+SW_SIGN_P=./sign/swsign-p.bin
+SW_SIGN_Q=./sign/swsign-q.bin
+SW_SIGN_R=./sign/swsign-r.bin
```
All these relative paths would break if this tool was installed.

Why wouldn't someone want to install this tool on a machine kept in a secure location, not network connected and use it to sign keys?

stewart-ibm commented on scripts/SignImage:
```diff
@@ -0,0 +1,114 @@
+
+
+# finally stitch the container header to the image 
+cat ${CONT_FILE} ${IMG_FILE} > ${CONT_IMG_FILE}
```
Why can't the signtool itself do all of this?

stewart-ibm commented on src/signtool/IBM_Container.cpp:
```diff
@@ -0,0 +1,1767 @@
+        }
+    };
+
+    uint16_t getUint16( const uint8_t *data )
```
get it from what? In what format?

Why are you writing custom endian routines? What exactly is wrong with hbe16toh(3) in endian.h ?

stewart-ibm commented on src/signtool/IBM_Container.cpp:
```diff
@@ -0,0 +1,1767 @@
+                                  ECDSA521_KEY_SIZE + 1 ); // since keyfile is 133 bytes with 0x04
+                                                           // at the begining
+
+            if (p_buffer[0] != g_UNCOMPRESSED_PUBKEY_FORMAT)
```
Check for valid length of file? What if it's a 0 byte length file?

stewart-ibm commented on src/signtool/IBM_Container.cpp:
```diff
@@ -0,0 +1,1767 @@
+                ss << "File <" 
+                   << p_keyFileName 
+                   << "> is not a vaid p521 public key file"
+                   << std::endl;
```
Assuming this is an error that would be printed to a user, the way you're constructing that string means you cannot translate it.

stewart-ibm commented on src/signtool/IBM_Container.cpp:
```diff
@@ -0,0 +1,1767 @@
+/**
+ *  @brief   print the Container Header
+ */
+void ContainerHdr::PrintHeader() const
```
Why not do the << magic and then you're not forced to print to cout, but can instead just redirect to any ostream ?

stewart-ibm commented on src/signtool/IBM_Container.cpp:
```diff
@@ -0,0 +1,1767 @@
+
+    padLen -= hdrBytes.size();
+
+    // write the Prefix Header bytes
```
All of these kind of comments are superfluous.

If the code isn't *obviously* doing what's written in this comment, then the code needs fixing, not commenting.

stewart-ibm commented on src/signtool/Makefile.am:
```diff
@@ -0,0 +1,48 @@
+        IBM_Exception.cpp \
+        IBM_Utils.cpp  \
+        IBM_Crypto.cpp  \
+        signtool.cpp
```
You're missing noinst_HEADERS, so `make distcheck` doesn't work:

```
  CXX      signtool-IBM_Container.o
../../../../src/signtool/IBM_Container.cpp:36:24: fatal error: IBM_Crypto.h: No such file or directory
```

stewart-ibm commented on src/signtool/IBM_Exception.h:
```diff
@@ -0,0 +1,59 @@
+#include <string>
+#include <stdexcept>
+
+class IBM_Exception : public std::exception
```
Why?

Why does IBM_Exception even exist? IBM isn't throwing an exception, it's not a special IBM problem that's occured.

stewart-ibm commented on src/signtool/IBM_HexString.h:
```diff
@@ -0,0 +1,194 @@
+
+typedef std::vector<byte> IBM_HexBytes;
+
+class IBM_HexString
```
Why is this some kind of special IBM style of hexstring?

IS it like powerpc bit ordering, in that we have to be all special?

stewart-ibm commented on src/signtool/IBM_Utils.h:
```diff
@@ -0,0 +1,92 @@
+};
+
+
+class IBM_Utils
```
catch all classes are seldom a good sign.

stewart-ibm commented on src/signtool/IBM_Utils.h:
```diff
@@ -0,0 +1,92 @@
+ 
+    ~IBM_Utils(); 
+    
+    static IBM_Utils* get(); 
```
get what? an IBM_Utils? WHy would the new operator not work? WHy would you have more than one?

stewart-ibm commented on src/signtool/signtool.cpp:
```diff
@@ -0,0 +1,593 @@
+#include <string>
+#include <algorithm>
+
+#include "IBM_Crypto.h"
```
special kind of IBM crypto?

stewart-ibm commented on src/signtool/signtool.cpp:
```diff
@@ -0,0 +1,593 @@
+    static std::string s_hashAlgo = "sha512";
+
+    // assume sign_agent running locally by default
+    static std::string s_saHostName = "127.0.0.1";
```
I thought the Notes stuff was out?

stewart-ibm commented on src/signtool/signtool.cpp:
```diff
@@ -0,0 +1,593 @@
+        return s_PROGRAM_NAME.c_str();
+    }
+
+    static std::string  s_VERSION = "1.0.0";
```
should come from config.h

subras-ibm committed:
Fixed various review comments received
   * Removed all reference of Lotus notes from README.md
   * Added support to detect support for c++11/c++0x in configure.ac
   * changed mode from string to enum
   * deleted copy constructor and assignment operators
   * Initialized member variables to default values and removed constructors
   * Map initialzed using compile-time initializer
   * all constants now use constexpr

ghost commented:
So, there's also an existing sbsigntool (https://git.kernel.org/cgit/linux/kernel/git/jejb/sbsigntools.git/, packaged for ubuntu: https://launchpad.net/ubuntu/+source/sbsigntool ).

sbsigntool: utility for signing and verifying files for UEFI Secure Boot

So sb-signing-utils is pretty confusing with that.

jk-ozlabs commented:
It'd be super cool if we could just add support for the new formats to an existing tool...

ghost commented:
Jeremy Kerr notifications@github.com writes:

> It'd be super cool if we could just add support for the new formats to
> an existing tool...

Very much agree.

## 

Stewart Smith
OPAL Architect, IBM.

ghost commented:
cclaudio notifications@github.com writes:

> The op-signtool name seems to be something specific for
> openpower. IIRC, signtool tool will be used for openpower and powervm.

I'm sure folks doing PowerVM will be able to cope, and they're also a
_much_ smaller group than those who will be involved in signing
OpenPOWER firmware.

## 

Stewart Smith
OPAL Architect, IBM.

subras-ibm committed:
More review comments and added new command get public key.

stewart-ibm commented on src/signtool/IBM_Utils.cpp:
```diff
@@ -257,3 +257,34 @@ void IBM_Utils::GetSignatureBytes( const std::string& p_fileName,
+              ((uint64_t)data[1] << 48) | ((uint64_t)data[0] << 56) );
+
+    return value;
+}
```
again, *WHY* are these here and you're not using normal endian routines? Are these somehow special or superior?

stewart-ibm commented on src/signtool/Makefile.am:
```diff
@@ -37,7 +39,7 @@ signtool_SOURCES = \
 CXXCPP0X=$(if ifeq @HAVE_CXX11@ "0",-std=c++0x,)
 
-signtool_CPPFLAGS = $(CXXCPP) $(CXXCPP0X) -I. -I -g3
+signtool_CPPFLAGS = $(CXXCPP) $(CXXCPP0X) -I. -I -g3 
```
why this change?

Trailing whitespace--

stewart-ibm commented on src/signtool/IBM_Socket.h:
```diff
@@ -0,0 +1,70 @@
+
+private:
+
+    int                  m_socket;      //!< handle to socket
```
It's not a handle, this isn't Win32.

stewart-ibm commented on src/signtool/IBM_Socket.h:
```diff
@@ -0,0 +1,70 @@
+#ifndef IBM_Socket_h_
```
No (C) header

stewart-ibm commented on src/signtool/IBM_Socket.h:
```diff
@@ -0,0 +1,70 @@
+ *
+ * A general interface to a socket 
+ */
+class IBM_Socket
```
Why does this still exist? No sockets to Notes.

stewart-ibm commented on src/signtool/IBM_Socket.h:
```diff
@@ -0,0 +1,70 @@
+        UDP_CLIENT,     //!< UDP client socket
+        TCP_SERVER,     //!< TCP Server socket
+        TCP_CLIENT      //!< TCP client socket
+    };
```
AF_UNIX, AF_INET, AF_INET6. These already exist, don't invent your own.

stewart-ibm commented on src/signtool/IBM_Socket.cpp:
```diff
@@ -0,0 +1,516 @@
+#include <sys/types.h>
```
No (C) header.

This code has enough problems that I'm not even going to go through it. I've gone over it before.

ghost commented:
Just to be clear: still a NAK from me.

subras-ibm committed:
Added helper perl script to build container

ghost commented:
<pre>
=================================================================
==23283==ERROR: LeakSanitizer: detected memory leaks
Direct leak of 56 byte(s) in 1 object(s) allocated from:
    #0 0x7f99e5cdae60 in malloc (/lib64/libasan.so.3+0xc6e60)
    #1 0x7f99e58220b7 in CRYPTO_malloc (/lib64/libcrypto.so.10+0x6e0b7)
Indirect leak of 1284 byte(s) in 18 object(s) allocated from:
    #0 0x7f99e5cdae60 in malloc (/lib64/libasan.so.3+0xc6e60)
    #1 0x7f99e58220b7 in CRYPTO_malloc (/lib64/libcrypto.so.10+0x6e0b7)
SUMMARY: AddressSanitizer: 1340 byte(s) leaked in 19 allocation(s).
</pre>

ghost commented:
<pre>
 /home/stewart/sb-signing-utils/src/signtool/signtool --mode development --imagefile zImage.epapr.container  --fldname sw-keyp --fldvalue /tmp/lrtVvmEgFK/swpubkeyp
==13039== Memcheck, a memory error detector
==13039== Copyright (C) 2002-2015, and GNU GPL'd, by Julian Seward et al.
==13039== Using Valgrind-3.11.0 and LibVEX; rerun with -h for copyright info
==13039== Command: /home/stewart/sb-signing-utils/src/signtool/signtool --mode development --imagefile zImage.epapr.container --fldname sw-keyp --fldvalue /tmp/lrtVvmEgFK/swpubkeyp
==13039== 
==13039== 
==13039== HEAP SUMMARY:
==13039==     in use at exit: 1,756 bytes in 25 blocks
==13039==   total heap usage: 230 allocs, 205 frees, 113,334 bytes allocated
==13039== 
==13039== 1,340 (56 direct, 1,284 indirect) bytes in 1 blocks are definitely lost in loss record 25 of 25
==13039==    at 0x4C2BBAD: malloc (vg_replace_malloc.c:299)
==13039==    by 0x4EA60B7: CRYPTO_malloc (in /usr/lib64/libcrypto.so.1.0.2h)
==13039==    by 0x4F20849: EC_KEY_new (in /usr/lib64/libcrypto.so.1.0.2h)
==13039==    by 0x4F216D1: ??? (in /usr/lib64/libcrypto.so.1.0.2h)
==13039==    by 0x4F22AC0: ??? (in /usr/lib64/libcrypto.so.1.0.2h)
==13039==    by 0x4F7C112: X509_PUBKEY_get (in /usr/lib64/libcrypto.so.1.0.2h)
==13039==    by 0x4F7C2C0: d2i_PUBKEY (in /usr/lib64/libcrypto.so.1.0.2h)
==13039==    by 0x4F93BAD: PEM_ASN1_read_bio (in /usr/lib64/libcrypto.so.1.0.2h)
==13039==    by 0x4F90E84: PEM_ASN1_read (in /usr/lib64/libcrypto.so.1.0.2h)
==13039==    by 0x414B72: IBM_Utils::GetPublicKeyBytes(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::vector<unsigned char, std::allocator<unsigned char> >&) (IBM_Utils.cpp:169)
==13039==    by 0x409F02: (anonymous namespace)::ReadPublicKeyFromFile(IBM_Mode, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::vector<unsigned char, std::allocator<unsigned char> >&) [clone .constprop.210] (IBM_Container.cpp:147)
==13039==    by 0x40B2C7: IBM_Container::UpdateField(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >) (IBM_Container.cpp:1213)
==13039== 
==13039== LEAK SUMMARY:
==13039==    definitely lost: 56 bytes in 1 blocks
==13039==    indirectly lost: 1,284 bytes in 18 blocks
==13039==      possibly lost: 0 bytes in 0 blocks
==13039==    still reachable: 416 bytes in 6 blocks
==13039==         suppressed: 0 bytes in 0 blocks
==13039== Reachable blocks (those to which a pointer was found) are not shown.
==13039== To see them, rerun with: --leak-check=full --show-leak-kinds=all
==13039== 
==13039== For counts of detected and suppressed errors, rerun with: -v
==13039== ERROR SUMMARY: 1 errors from 1 contexts (suppressed: 0 from 0)
</pre>

ghost commented:
<pre>
==13041== Memcheck, a memory error detector
==13041== Copyright (C) 2002-2015, and GNU GPL'd, by Julian Seward et al.
==13041== Using Valgrind-3.11.0 and LibVEX; rerun with -h for copyright info
==13041== Command: /home/stewart/sb-signing-utils/src/signtool/signtool --mode development --sign --projname /home/stewart/op-build/openpower/package/sb-signing-utils/keys/hw_key_a.key --sigfile /tmp/lrtVvmEgFK/hwsignA.bin --digest 065B5D5DF63E74F1A5EA7ABAF9990788C8DB5B7BCE37F46559CB47AD888D989723F556747E2798FE249BEA2299E1A4BCD8437748466504A62DB8904F0AA63C7D
==13041== 
ECC Key type: UNDEF
signature :30818702410AAE767A9618BA2605B4905A094FE37FF38FBBD840E0DEC7CF8AE26391C40A3C5228D0D1E8DF7421A7CC0FD2806E3CC869789290CD5BE0BBFD7A85
45345ED4EE53024201C1DB5769F793CA6EC31CA73E1BA12E6A2E9ACD575A37C0682946AF407281325E49A513DEC6BCB7EE3192E9F2DBB3BCEEE60C366F2338BB
B9461C110B8822EA5E6A
==13041== 
==13041== HEAP SUMMARY:
==13041==     in use at exit: 2,556 bytes in 38 blocks
==13041==   total heap usage: 3,215 allocs, 3,177 frees, 226,114 bytes allocated
==13041== 
==13041== 2,028 (56 direct, 1,972 indirect) bytes in 1 blocks are definitely lost in loss record 38 of 38
==13041==    at 0x4C2BBAD: malloc (vg_replace_malloc.c:299)
==13041==    by 0x4EA60B7: CRYPTO_malloc (in /usr/lib64/libcrypto.so.1.0.2h)
==13041==    by 0x4F20849: EC_KEY_new (in /usr/lib64/libcrypto.so.1.0.2h)
==13041==    by 0x4F1FF04: d2i_ECPrivateKey (in /usr/lib64/libcrypto.so.1.0.2h)
==13041==    by 0x4F22B7D: ??? (in /usr/lib64/libcrypto.so.1.0.2h)
==13041==    by 0x4F7EDE7: d2i_PrivateKey (in /usr/lib64/libcrypto.so.1.0.2h)
==13041==    by 0x4F944AE: PEM_read_bio_PrivateKey (in /usr/lib64/libcrypto.so.1.0.2h)
==13041==    by 0x4F948CB: PEM_read_PrivateKey (in /usr/lib64/libcrypto.so.1.0.2h)
==13041==    by 0x418D70: IBM_Crypto::doOpensslSign(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::vector<unsigned char, std::allocator<unsigned char> > const&, std::vector<unsigned char, std::allocator<unsigned char> >&) (IBM_Crypto.cpp:556)
==13041==    by 0x417D22: IBM_Crypto::Sign(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, unsigned short) (IBM_Crypto.cpp:213)
==13041==    by 0x404CFC: main (signtool.cpp:519)
==13041== 
==13041== LEAK SUMMARY:
==13041==    definitely lost: 56 bytes in 1 blocks
==13041==    indirectly lost: 1,972 bytes in 27 blocks
==13041==      possibly lost: 0 bytes in 0 blocks
==13041==    still reachable: 528 bytes in 10 blocks
==13041==         suppressed: 0 bytes in 0 blocks
==13041== Reachable blocks (those to which a pointer was found) are not shown.
==13041== To see them, rerun with: --leak-check=full --show-leak-kinds=all
==13041== 
==13041== For counts of detected and suppressed errors, rerun with: -v
==13041== ERROR SUMMARY: 1 errors from 1 contexts (suppressed: 0 from 0)
</pre>

stewart-ibm commented on src/signtool/IBM_Container.cpp:
```diff
@@ -0,0 +1,1712 @@
+        packet.insert( packet.end(),
+                       m_swPkeyR,
+                       (m_swPkeyR + sizeof(m_swPkeyR)) );
+    }
```
Correction, that is the data structure, it's just not actually bloody documented anywhere, like, say, in the headers describing the container format.

subras-ibm committed:
1. Fixed memory leak due to not releasing memory allocated by getaddrinfo() call
2. Changed all delete functions as public, so compiler can give better error messages

subras-ibm committed:
1. Added container layout description in header file
2. Fixed memory leak, released OpenSSL allocated memory
3. Fixed THROW_EXCEPTION macro usage in signtool.cpp

subras-ibm committed:
1. Added command to also get public key bytes from a given private key file for
   development mode

Commit:
Fixed condition check

subras-ibm committed:
Fixed signature verification failure in certain cases, where r & s field is 64 bytes
long, this case was not handled correctly.

subras-ibm committed:
Merge branch 'master' of https://github.com/subras-ibm/sb-signing-utils

subras-ibm committed:
Added support for config file to specify, signing agent host and port number,
project details to obtain project name, public key file and sign file by
specifying a token.

subras-ibm committed:
Modified config file format to specify project name, pubkey and signature
as one entry

subras-ibm committed:
Fixed swap of sign and pubkey file parameters, added missing commands and parameters in usage

subras-ibm committed:
Added config file support as discussed
    1. Support for [DEFAULTS] section
    2. Create sign filename as
          <token> + "_" + <key> + ".sign"
    3. Create pubkey filename as
          <token> + "_" + <key> + ".pub"
    4. Use values from DEFAULTS section for any missing keys

subras-ibm committed:
Create sign and file names as
  1. if type is HW then filenames are
       <token> + "_" + [a|b|c].sign
       <token> + "_" + [a|b|c].pub
  2. if type is SW then filenames are
       <token> + "_" + [p|q|r].sign
       <token> + "_" + [p|q|r].pub

subras-ibm committed:
Fix verify command when using configfile option

subras-ibm committed:
Added support for mode in config file and renamed mode "ibm-production" to more
generic mode name "production".

Commit:
Add support for production-mode signing to crtSignedContainer.pl

Commit:
fixed stray character in the prolog

subras-ibm committed:
Merge pull request #1 from eggler/master

Add support for production-mode signing to crtSignedContainer.pl

ghost commented:
Last year, I tried to use this 8,500 lines of code to simply wrap a skiboot.lid with enough of a secure boot header to boot with a trusted boot enabled hostboot (**not** secure boot) and I couldn't work it out. It seemed like I needed about 20 invocations of the utility to get anywhere near producing a valid secure boot header.

So I wrote my own: https://github.com/open-power/skiboot/blob/master/libstb/create-container.c

It's 118 lines of C and it wraps skiboot.lid in enough of a header for hostboot to boot it.

Considering that the OpenSSL library calls required to produce the signatures look to be about a dozen or so lines of code (add in some command line parsing etc much like sbsign for UEFI systems) and we're still looking at *maybe* 500 lines of C. This is *less* than simply the *perl* script in this pull request.

Considering that pretty much all of the review comments have *still* not been addressed and we're now **four** months in...  I'm thinking the better avenue to go down will be to extend the create-container code I wrote, possibly just adding our code into sbsigntool.

I had hoped that this code would have come around to something I'd feel okay in having as part of op-build, and we could be getting all nice and ready for secure/trusted boot everywhere in the future.

But it's not ready and there's no sign of improvement... so we may as well close this pull request, mark the project as unused and pursue other ways of generating the container.

ghost commented:
Superseded by https://github.com/open-power/sb-signing-utils/pull/4
